backends/tfhe-cuda-backend/cuda/src/polynomial/functions.cuh

```diff
@@ -191,67 +191,86 @@ __device__ void add_to_torus(double2 *m_values, Torus *result,
-      tid = tid + params::degree / params::opt;
-    }
-    synchronize_threads_in_block();
-
-    // Set ACC = -ACC
-    tid = threadIdx.x;
-#pragma unroll
-    for (int i = 0; i < params::opt; i++) {
-      accumulator_slice[tid] =
-          SEL(-result[i], result[i], tid >= params::degree - nth);
+      auto x = glwe_slice[params::degree - tid - 1];
+      result[i] = SEL(-x, x, tid >= params::degree - nth);
       tid = tid + params::degree / params::opt;
     }
     synchronize_threads_in_block();
 
     // Perform ACC * X
     // (equivalent to multiply_by_monomial_negacyclic_inplace(1))
+    // Copy to the mask of the LWE sample
     tid = threadIdx.x;
-    result[params::opt];
     for (int i = 0; i < params::opt; i++) {
       // if (tid < 1)
-      //  result[i] = -accumulator_slice[tid - 1 + params::degree];
+      //  result[i] = -glwe_slice[tid - 1 + params::degree];
       // else
-      //  result[i] = accumulator_slice[tid - 1];
-      int x = tid - 1 + SEL(0, params::degree - nth, tid < 1);
-      result[i] = SEL(1, -1, tid < 1) * accumulator_slice[x];
-      tid += params::degree / params::opt;
-    }
-    synchronize_threads_in_block();
+      //  result[i] = glwe_slice[tid - 1];
+      uint32_t dst_idx = tid + 1 + nth;
+      if (dst_idx == params::degree)
+        lwe_array_out_slice[0] = -result[i];
+      else {
+        dst_idx =
+            SEL(dst_idx, dst_idx - params::degree, dst_idx >= params::degree);
+        lwe_array_out_slice[dst_idx] = result[i];
+      }
 
-    // Copy to the mask of the LWE sample
-    tid = threadIdx.x;
-#pragma unroll
-    for (int i = 0; i < params::opt; i++) {
-      lwe_array_out_slice[tid] = result[i];
-      tid = tid + params::degree / params::opt;
+      tid += params::degree / params::opt;
     }
   }
 }
 
+template <typename Torus, class params>
+__global__ void apply_sample_extract(Torus *lwe_array_out, Torus *glwe_in,
+                                     uint32_t *nth_array,
+                                     uint32_t glwe_dimension) {
+
+  const int input_id = blockIdx.x;
+
+  const int lwe_output_size = glwe_dimension * params::degree + 1;
+
+  auto lwe_out = lwe_array_out + input_id * lwe_output_size;
+  auto nth = nth_array[input_id];
+
+  sample_extract_mask<Torus, params>(lwe_out, glwe_in, glwe_dimension, nth);
+  sample_extract_body<Torus, params>(lwe_out, glwe_in, glwe_dimension, nth);
+}
+
+template <typename Torus, class params>
+__host__ void host_sample_extract(cudaStream_t *streams, Torus *lwe_array_out,
+                                  Torus *glwe_in, uint32_t *nth_array,
+                                  uint32_t num_samples,
+                                  uint32_t glwe_dimension) {
+
+  dim3 grid(num_samples);
+  dim3 thds(params::degree / params::opt);
+  apply_sample_extract<Torus, params><<<grid, thds, 0, streams[0]>>>(
+      lwe_array_out, glwe_in, nth_array, glwe_dimension);
+  check_cuda_error(cudaGetLastError());
+}
+
```

Comment:
Hmm ok this was in functions.cuh already. Well what do you think of moving it to ciphertext.cuh?

Reply:
I think it's a great idea. Sample extraction relates directly to GLWE and LWE ciphertexts.

Reply:
Done.